// printit/main.js
// Generated by CoffeeScript 1.9.0
var Logger, colors, dateFormat, levelColors,
  __slice = [].slice;

dateFormat = require('./date-format');

module.exports = function(options) {
  return new Logger(options);
};

colors = {
  blue: ['\x1B[34m', '\x1B[39m'],
  cyan: ['\x1B[36m', '\x1B[39m'],
  green: ['\x1B[32m', '\x1B[39m'],
  magenta: ['\x1B[36m', '\x1B[39m'],
  red: ['\x1B[31m', '\x1B[39m'],
  yellow: ['\x1B[33m', '\x1B[39m']
};

levelColors = {
  error: colors.red,
  debug: colors.green,
  warn: colors.yellow,
  info: colors.blue
};

Logger = (function() {
  function Logger(_at_options) {
    this.options = _at_options;
    if (this.options == null) {
      this.options = {};
    }
    if (this.options.date && (this.options.dateFormat == null)) {
      this.options.dateFormat = 'YYYY-MM-DD hh:mm:ss:S';
    }
  }

  Logger.prototype.colorify = function(text, color) {
    return "" + color[0] + text + color[1];
  };

  Logger.prototype.stringify = function(text) {
    if (text instanceof Object) {
      text = JSON.stringify(text);
    }
    return text;
  };

  Logger.prototype.format = function(level, texts) {
    var date, text;
    text = ((function() {
      var _i, _len, _results;
      _results = [];
      for (_i = 0, _len = texts.length; _i < _len; _i++) {
        text = texts[_i];
        _results.push(this.stringify(text));
      }
      return _results;
    }).call(this)).join(" ");
    if (this.options.prefix != null) {
      text = this.options.prefix + " | " + text;
    }
    if (process.env.NODE_ENV !== 'production') {
      level = this.colorify(level, levelColors[level]);
    }
    if (level) {
      text = level + " - " + text;
    }
    if (this.options.date) {
      date = new Date().format(this.options.dateFormat);
      text = "[" + date + "] " + text;
    }
    return text;
  };

  Logger.prototype.info = function() {
    var texts;
    texts = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
    if (process.env.NODE_ENV !== 'test') {
      return console.info(this.format('info', texts));
    }
  };

  Logger.prototype.warn = function() {
    var texts;
    texts = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
    if (process.env.NODE_ENV !== 'test') {
      return console.warn(this.format('warn', texts));
    }
  };

  Logger.prototype.error = function() {
    var errDetails, filePath, line, s, sp, stackReg, stackReg2, stacklist, texts;
    texts = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
    stacklist = (new Error()).stack.split('\n').slice(2);
    stackReg = /at\s+(.*)\s+\((.*):(\d*):(\d*)\)/gi;
    stackReg2 = /at\s+()(.*):(\d*):(\d*)/gi;
    s = stacklist[0];
    sp = stackReg.exec(s) || stackReg2.exec(s);
    filePath = sp[2].substr(process.cwd().length);
    line = sp[3];
    errDetails = "." + filePath + ":" + line + " |";
    texts.unshift(errDetails);
    if (process.env.NODE_ENV !== 'test') {
      return console.error(this.format('error', texts));
    }
  };

  Logger.prototype.debug = function() {
    var texts;
    texts = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
    if (process.env.DEBUG) {
      return console.info(this.format('debug', texts));
    }
  };

  Logger.prototype.raw = function() {
    var texts;
    texts = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
    return console.log.apply(console, texts);
  };

  Logger.prototype.lineBreak = function(text) {
    return this.raw(Array(80).join("*"));
  };

  return Logger;

})();
